codec-http2/src/test/java/io/netty/handler/codec/http2/StreamBufferingEncoderTest.java

```diff
@@ -0,0 +1,413 @@
+        writeVerifyWriteHeaders(never(), 7, promise);
+
+        encoder.writeRstStream(ctx, 3, CANCEL.code(), promise);
+        assertEquals(1, connection.numActiveStreams());
+        assertEquals(1, encoder.numBufferedStreams());
+        encoder.writeRstStream(ctx, 5, CANCEL.code(), promise);
+        assertEquals(1, connection.numActiveStreams());
+        assertEquals(0, encoder.numBufferedStreams());
+        encoder.writeRstStream(ctx, 7, CANCEL.code(), promise);
+        assertEquals(0, connection.numActiveStreams());
+        assertEquals(0, encoder.numBufferedStreams());
+    }
+
+    @Test
+    public void bufferUntilMaxStreamsIncreased() {
+        encoder.writeSettingsAck(ctx, promise);
+        setMaxConcurrentStreams(2);
+
+        encoderWriteHeaders(3, promise);
+        encoderWriteHeaders(5, promise);
+        encoderWriteHeaders(7, promise);
+        encoderWriteHeaders(9, promise);
+        assertEquals(2, encoder.numBufferedStreams());
+
+        writeVerifyWriteHeaders(times(1), 3, promise);
+        writeVerifyWriteHeaders(times(1), 5, promise);
+        writeVerifyWriteHeaders(never(), 7, promise);
+        writeVerifyWriteHeaders(never(), 9, promise);
+
+        // Simulate that we received a SETTINGS frame which
+        // increased MAX_CONCURRENT_STREAMS to 5.
+        setMaxConcurrentStreams(5);
+        encoder.writeSettingsAck(ctx, promise);
+
+        assertEquals(0, encoder.numBufferedStreams());
+        writeVerifyWriteHeaders(times(1), 7, promise);
+        writeVerifyWriteHeaders(times(1), 9, promise);
+
+        encoderWriteHeaders(11, promise);
+
+        writeVerifyWriteHeaders(times(1), 11, promise);
+
+        assertEquals(5, connection.local().numActiveStreams());
+    }
+
+    @Test
+    public void bufferUntilSettingsReceived() throws Http2Exception {
+        int initialLimit = SMALLEST_MAX_CONCURRENT_STREAMS;
+        int numStreams = initialLimit * 2;
+        for (int ix = 0, nextStreamId = 3; ix < numStreams; ++ix, nextStreamId += 2) {
+            encoderWriteHeaders(nextStreamId, promise);
+            if (ix < initialLimit) {
+                writeVerifyWriteHeaders(times(1), nextStreamId, promise);
+            } else {
+                writeVerifyWriteHeaders(never(), nextStreamId, promise);
+            }
+        }
+        assertEquals(numStreams / 2, encoder.numBufferedStreams());
+
+        // Simulate that we received a SETTINGS frame.
+        setMaxConcurrentStreams(initialLimit * 2);
+
+        assertEquals(0, encoder.numBufferedStreams());
+        assertEquals(numStreams, connection.local().numActiveStreams());
+    }
+
+    @Test
+    public void exhaustedStreamsDoNotBuffer() throws Http2Exception {
+        // Write the highest possible stream ID for the client.
+        // This will cause the next stream ID to be negative.
+        encoderWriteHeaders(Integer.MAX_VALUE, promise);
+
+        // Disallow any further streams.
+        setMaxConcurrentStreams(0);
```

Comment:
Do we have to have this?  Seems like we should fail regardless of this condition?